webapp/app/routes/app/emergency/$id/_route.tsx

```diff
@@ -0,0 +1,239 @@
+import { displayBytes32 } from "@/routes/app/proposals/$id/common-tables";
+import SignButton from "@/routes/app/proposals/$id/sign-button";
+import { requireUserFromHeader } from "@/utils/auth-headers";
+import { badRequest, notFound } from "@/utils/http";
+import { type ActionFunctionArgs, type LoaderFunctionArgs, json } from "@remix-run/node";
+import { useLoaderData, useNavigate } from "@remix-run/react";
+import { ArrowLeft } from "lucide-react";
+import { getFormData, getParams } from "remix-params-helper";
+import { $path } from "remix-routes";
+import { zodHex } from "validate-cli";
+import { type Hex, isAddressEqual } from "viem";
+import { z } from "zod";
+
+export async function loader(args: LoaderFunctionArgs) {
+  const user = requireUserFromHeader(args.request);
+  const params = getParams(args.params, z.object({ id: zodHex }));
+  if (!params.success) {
+    throw notFound();
+  }
+
+  const { id: proposalId } = params.data;
+
+  const boardAddress = await emergencyBoardAddress();
+
+  const proposal = await getEmergencyProposalByExternalId(proposalId);
+
+  if (proposal === undefined) {
+    throw notFound();
+  }
+
+  const signatures = await getSignaturesByEmergencyProposalId(proposal.externalId);
+
+  return json({
+    proposal: {
+      title: proposal?.title,
+      externalId: proposal.externalId,
+      proposedOn: proposal.proposedOn,
+      calldata: proposal.calldata,
+      targetAddress: proposal.targetAddress,
+      salt: proposal.salt,
+      value: proposal.value.toString(),
+    },
+    addresses: {
+      emergencyBoard: boardAddress,
+      zkFoundation: await zkFoundationAddress(),
+    },
+    user,
+    signatures,
+    allGuardians: await guardianMembers(),
+    allSecurityCouncil: await councilMembers(),
+  });
+}
+
+export async function action({ request }: ActionFunctionArgs) {
+  const user = requireUserFromHeader(request);
+  const formData = await getFormData(
+    request,
+    z.object({
+      signature: zodHex,
+      proposalId: zodHex,
+      actionName: actionSchema,
+    })
+  );
+  if (!formData.success) {
+    throw badRequest("Failed to parse signature data");
+  }
+  const data = formData.data;
+
+  await saveEmergencySignature(
+    data.signature,
+    user.address as Hex,
+    data.actionName,
+    data.proposalId
+  );
+
+  return json({ ok: true });
+}
+
+const ACTION_NAMES = {
+  guardian: "ExecuteEmergencyUpgradeGuardians",
+  securityCouncil: "ExecuteEmergencyUpgradeSecurityCouncil",
+  zkFoundation: "ExecuteEmergencyUpgradeZKFoundation",
+};
+
+export default function EmergencyUpgradeDetails() {
+  const navigate = useNavigate();
+
+  const { user, proposal, addresses, signatures, allSecurityCouncil, allGuardians } =
+    useLoaderData<typeof loader>();
+
+  if (user.role === "visitor") {
+    return "Unauthorized: Only valid signers can see this page.";
+  }
+
+  const actionName = ACTION_NAMES[user.role];
+  const GUARDIAN_THRESHOLD = 5;
+  const SC_THRESHOLD = 9;
```

Comment:
Move magic thresholds to a system constants file, or if possible read it from smart contract if it's stored there.

e.g. 
```
const guardiansThreshold = await viemClient.readContract({
  address: guardiansAddress,
  abi: GUARDIANS_RAW_ABI
  functionName: 'APPROVE_UPGRADE_GUARDIANS_THRESHOLD'
})
```